File: sdks/clojure/test/cvisor/remote_smoke.clj
(ns cvisor.remote-smoke
  "Remote-GraphQL e2e smoke for the Clojure SDK's pure-JDK client
  (cvisor.remote over java.net.http — no FFM/libcvisor). Run against a running
  cvisord:

    CVISOR_GRAPHQL_URL=http://127.0.0.1:8080/graphql CVISOR_TOKEN=... clojure -M:remote"
  (:require [cvisor.remote :as remote])
  (:import [java.nio.charset StandardCharsets]))

(defn -main [& _]
  (let [url (or (System/getenv "CVISOR_GRAPHQL_URL") "http://127.0.0.1:8080/graphql")
        token (or (System/getenv "CVISOR_TOKEN") "")
        client (remote/client url token)]
    (assert (true? (get (remote/health client) "ok")) "health not ok")

    (let [out (remote/run client "echo hello")]
      (assert (= "hello\n" (get out "stdout")) (str "run stdout: " out))
      (assert (= 0 (get out "exitCode")) (str "run exit code: " out)))

    (let [sb (remote/create-sandbox client)
          id (get sb "id")]
      (assert (not (empty? id)) (str "create-sandbox returned no id: " sb))
      (remote/write-file client id "/tmp/data.txt" "round-trip\n")
      (let [data (String. ^bytes (remote/read-file client id "/tmp/data.txt")
                          StandardCharsets/UTF_8)]
        (assert (= "round-trip\n" data) (str "read-file round-trip: " data)))
      (remote/free-sandbox client id))

    (println "CLOJURE_GRAPHQL_OK")
    (flush)
    (System/exit 0)))

File: sdks/scala/src/main/scala/dev/tsirysndr/cvisor/RemoteSmoke.scala
package dev.tsirysndr.cvisor

import java.nio.charset.StandardCharsets

/** Remote-GraphQL e2e smoke for the Scala SDK's portable client
  * ([[RemoteSandbox]] over `java.net.http` — no libcvisor). Run against a
  * running cvisord:
  *
  * {{{
  * CVISOR_GRAPHQL_URL=http://127.0.0.1:8080/graphql CVISOR_TOKEN=... \
  *   sbt "runMain dev.tsirysndr.cvisor.RemoteSmoke"
  * }}}
  */
object RemoteSmoke:
  private def check(cond: Boolean, msg: => String): Unit =
    if !cond then
      System.err.println(s"assertion failed: $msg")
      sys.exit(1)

  def main(args: Array[String]): Unit =
    val url = sys.env
      .get("CVISOR_GRAPHQL_URL")
      .orElse(sys.env.get("CVISOR_URL"))
      .getOrElse("http://127.0.0.1:8080/graphql")
    val token = sys.env.getOrElse("CVISOR_TOKEN", "")
    val sb = RemoteSandbox(url, token)

    val health = sb.health()
    check(health.ok, s"health not ok: $health")

    val out = sb.run("echo hello")
    check(out.stdout == "hello\n", s"run stdout: $out")
    check(out.exitCode == 0, s"run exit code: $out")

    val info = sb.createSandbox()
    check(info.id.nonEmpty, s"createSandbox returned no id: $info")
    sb.writeFile("/tmp/data.txt", "round-trip\n")
    val data = new String(sb.readFile("/tmp/data.txt"), StandardCharsets.UTF_8)
    check(data == "round-trip\n", s"readFile round-trip: $data")
    sb.freeSandbox()

    println("SCALA_GRAPHQL_OK")
